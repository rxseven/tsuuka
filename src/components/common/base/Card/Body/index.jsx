import PropTypes from 'prop-types';
import exact from 'prop-types-exact';
import React from 'react';

const propTypes = exact({
  children: PropTypes.node.isRequired
});

function Body({ children }) {
  return <div className="card-body">{children}</div>;
}

Body.propTypes = propTypes;

export default Body;
