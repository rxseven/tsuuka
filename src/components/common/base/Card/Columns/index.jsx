import PropTypes from 'prop-types';
import exact from 'prop-types-exact';
import React from 'react';

const propTypes = exact({
  children: PropTypes.node.isRequired
});

function Columns({ children }) {
  return <div className="card-columns">{children}</div>;
}

Columns.propTypes = propTypes;

export default Columns;
