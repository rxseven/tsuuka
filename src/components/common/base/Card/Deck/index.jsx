import PropTypes from 'prop-types';
import exact from 'prop-types-exact';
import React from 'react';

const propTypes = exact({
  children: PropTypes.node.isRequired
});

function Deck({ children }) {
  return <div className="card-deck">{children}</div>;
}

Deck.propTypes = propTypes;

export default Deck;
