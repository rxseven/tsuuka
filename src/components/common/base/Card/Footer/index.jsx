import PropTypes from 'prop-types';
import exact from 'prop-types-exact';
import React from 'react';

const propTypes = exact({
  children: PropTypes.node.isRequired
});

function Footer({ children }) {
  return <div className="card-footer text-muted">{children}</div>;
}

Footer.propTypes = propTypes;

export default Footer;
