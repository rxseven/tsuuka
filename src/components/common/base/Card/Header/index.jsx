import PropTypes from 'prop-types';
import exact from 'prop-types-exact';
import React from 'react';

const propTypes = exact({
  children: PropTypes.node.isRequired
});

function Header({ children }) {
  return <div className="card-header">{children}</div>;
}

Header.propTypes = propTypes;

export default Header;
