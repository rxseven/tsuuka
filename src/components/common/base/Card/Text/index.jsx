import PropTypes from 'prop-types';
import exact from 'prop-types-exact';
import React from 'react';

const propTypes = exact({
  children: PropTypes.node.isRequired
});

function Text({ children }) {
  return <p className="card-text">{children}</p>;
}

Text.propTypes = propTypes;

export default Text;
