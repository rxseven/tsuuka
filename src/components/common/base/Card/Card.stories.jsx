import React from 'react';
import { storiesOf } from '@storybook/react';

import mock from 'stories/mock';
import Card from './index';

const { text } = mock;
const title = 'Components/Common/Base/Card';

storiesOf(title, module)
  .add('default', () => <Card>{text}</Card>)
  .add('alignment', () => <Card alignment="right">{text}</Card>)
  .add('background', () => (
    <Card background="primary" color="white">
      {text}
    </Card>
  ))
  .add('end', () => (
    <div>
      <Card end>{text}</Card>
      <p>Next element</p>
    </div>
  ))
  .add('margin-bottom', () => (
    <div>
      <Card margin="4">{text}</Card>
      <p>Next element</p>
    </div>
  ))
  .add('assembled', () => (
    <Card>
      <Card.Header>Header</Card.Header>
      <Card.Body>
        <Card.Title>Title</Card.Title>
        <Card.Text>{text}</Card.Text>
      </Card.Body>
      <Card.Footer>Footer</Card.Footer>
    </Card>
  ));
