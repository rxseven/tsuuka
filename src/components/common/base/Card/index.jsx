import cx from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';

import Body from './Body';
import Deck from './Deck';
import Columns from './Columns';
import Footer from './Footer';
import Header from './Header';
import Text from './Text';
import Title from './Title';

const propTypes = {
  alignment: PropTypes.string,
  background: PropTypes.string,
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  color: PropTypes.string,
  end: PropTypes.bool,
  margin: PropTypes.string
};

const defaultProps = {
  alignment: 'text-left',
  background: '',
  className: '',
  color: '',
  end: false,
  margin: '3'
};

function Card({
  alignment,
  background,
  children,
  className,
  color,
  end,
  margin
}) {
  return (
    <div
      className={cx(
        'card',
        className,
        !!alignment && `text-${alignment}`,
        !!background && `bg-${background}`,
        !!color && `text-${color}`,
        !end && `mb-${margin}`
      )}
    >
      {children}
    </div>
  );
}

Card.propTypes = propTypes;
Card.defaultProps = defaultProps;

Card.Body = Body;
Card.Columns = Columns;
Card.Deck = Deck;
Card.Footer = Footer;
Card.Header = Header;
Card.Text = Text;
Card.Title = Title;

export default Card;
