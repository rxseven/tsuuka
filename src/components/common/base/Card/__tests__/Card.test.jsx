import React from 'react';

import { factory, toMarkup } from 'tests/utilities';
import Card from '../index';

// Arrange
const seed = { content: 'content' };
const source = { children: <span>{seed.content}</span> };
const input = { ...seed, ...source };

// Setup
function setup(props) {
  return factory(Card, source, props);
}

// Test suites
describe('<Card />', () => {
  it('should render without crashing', () => {
    setup();
  });

  it('should render passed children correctly', () => {
    const expected = {
      content: input.content,
      html: input.children
    };
    const { component } = setup();

    expect(component).toHaveTextContent(expected.content);
    expect(component).toContainHTML(toMarkup(expected.html));
  });

  it('should align content to the left of the container by default', () => {
    const expected = { class: 'text-left' };
    const { component } = setup();

    expect(component).toHaveClass(expected.class);
  });

  it('should align content at the center of the container', () => {
    const props = { alignment: 'center' };
    const expected = { class: `text-${props.alignment}` };
    const { component } = setup(props);

    expect(component).toHaveClass(expected.class);
  });

  it('should have a "primary" background', () => {
    const props = { background: 'primary' };
    const expected = { class: `bg-${props.background}` };
    const { component } = setup(props);

    expect(component).toHaveClass(expected.class);
  });

  it('should render content with "red" color', () => {
    const props = { color: 'red' };
    const expected = { class: `text-${props.color}` };
    const { component } = setup(props);

    expect(component).toHaveClass(expected.class);
  });

  it('should render card with custom class names', () => {
    const props = { className: 'custom-class' };
    const expected = { class: props.className };
    const { component } = setup(props);

    expect(component).toHaveClass(expected.class);
  });

  it('should have "margin-bottom" with 1rem by default', () => {
    const expected = { class: 'mb-3' };
    const { component } = setup();

    expect(component).toHaveClass(expected.class);
  });

  it('should have "margin-bottom" with 1.5rem', () => {
    const props = { margin: '4' };
    const expected = { class: `mb-${props.margin}` };
    const { component } = setup(props);

    expect(component).toHaveClass(expected.class);
  });

  it('should not have "margin-bottom" at the end of the card', () => {
    const props = { end: true, margin: '3' };
    const expected = { class: props.margin };
    const { component } = setup(props);

    expect(component).not.toHaveClass(expected.class);
  });
});
