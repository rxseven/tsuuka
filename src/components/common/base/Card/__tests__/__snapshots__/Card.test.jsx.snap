// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> Snapshot tests should render correctly 1`] = `
<div
  className="card text-text-left mb-3"
>
  content
</div>
`;
